adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/sasl/AmqpAdapterSaslAuthenticatorFactory.java

```diff
@@ -0,0 +1,214 @@
+
+                    } else {
+                        LOG.debug("validation of credentials failed: " + outcome.cause().getMessage());
+                        sasl.done(SaslOutcome.PN_SASL_AUTH);
+
+                    }
+                    completionHandler.handle(Boolean.TRUE);
+                });
+
+                final byte[] saslResponse = new byte[sasl.pending()];
+                sasl.recv(saslResponse, 0, saslResponse.length);
+
+                verifyPlain(remoteMechanism, saslResponse, deviceAuthTracker.completer());
+            }
+        }
+
+        @Override
+        public boolean succeeded() {
+            return succeeded;
+        }
+
+        private void verifyPlain(final String remoteMechanism, final byte[] saslResponse,
+                final Handler<AsyncResult<Device>> completer) {
+
+            getFields(saslResponse).compose(fields -> {
+                final DeviceCredentials credentials = UsernamePasswordCredentials
+                        .create(fields[0], fields[1], config.isSingleTenant());
+                if (credentials == null) {
+                    // adapter is configured for multi-tenancy but credentials
+                    // does not comply with the pattern [<authId>@<tenantId>]
+                    completer.handle(Future.failedFuture(new CredentialException(
+                            "username does not comply with expected pattern [<authId>@<tenantId>]")));
+                } else {
+                    authProvider.authenticate(credentials, completer);
+                }
+                return Future.future();
+            }).recover(t -> {
+                completer.handle(Future.failedFuture(t));
+                return Future.failedFuture(t);
+            });
+
+        }
+
+        private Future<String[]> getFields(final byte[] saslResponse) {
+            final Future<String[]> result = Future.future();
+            final int authzidEndPosition = findNullByte(saslResponse, 0);
+            if (authzidEndPosition < 0) {
+                result.fail(
+                        new CredentialException("PLAIN response is malformed (does not contain an Authorization ID"));
+            }
+            final int authcidEndPosition = findNullByte(saslResponse, authzidEndPosition + 1);
+            if (authcidEndPosition < 0) {
+                result.fail(new CredentialException("PLAIN response must contain an authentication ID"));
+            }
+            if (authcidEndPosition == saslResponse.length - 1) {
+                result.fail(new CredentialException("PLAIN response must contain a password"));
+            }
+            // Get the authentication identity and password (ignoring authorization id)
+            final String authcid = new String(saslResponse, authzidEndPosition + 1,
+                    authcidEndPosition - (authzidEndPosition + 1), StandardCharsets.UTF_8);
+            final String passwd = new String(saslResponse, authcidEndPosition + 1,
```

Comment:
which charset is used here?

Reply:
good catch! should also be `StandardCharsets.UTF_8`
> Anyway, I do not think that we need multiple implementations of parsing the SASL response.

yes, that's true. i will drop this and use the implementation present with the `AuthenticationService`. can i move it to `AuthenticationConstants`?